mempool/src/shared_mempool.rs

```diff
@@ -277,92 +284,80 @@ where
 }
 
 /// This task handles inbound network events.
-async fn inbound_network_task<V>(smp: SharedMempool<V>, network_events: MempoolNetworkEvents)
-where
+async fn inbound_network_task<V>(
+    smp: SharedMempool<V>,
+    executor: TaskExecutor,
+    mut network_events: MempoolNetworkEvents,
+) where
     V: TransactionValidation,
 {
     let peer_info = smp.peer_info.clone();
     let subscribers = smp.subscribers.clone();
-    let max_inbound_syncs = smp.config.shared_mempool_max_concurrent_inbound_syncs;
-
-    // Handle the NewPeer/LostPeer events immediatedly, since they are not async
-    // and we don't want to buffer them or let them get reordered. The inbound
-    // direct-send messages are placed in a bounded FuturesUnordered queue and
-    // allowed to execute concurrently. The .buffer_unordered() also correctly
-    // handles back-pressure, so if mempool is slow the back-pressure will
-    // propagate down to network.
-    let f_inbound_network_task = network_events
-        .filter_map(move |network_event| {
-            trace!("SharedMempoolEvent::NetworkEvent::{:?}", network_event);
-            match network_event {
-                Ok(network_event) => match network_event {
-                    Event::NewPeer(peer_id) => {
-                        OP_COUNTERS.inc("smp.event.new_peer");
-                        new_peer(&peer_info, peer_id);
-                        notify_subscribers(
-                            SharedMempoolNotification::PeerStateChange,
-                            &subscribers,
-                        );
-                        future::ready(None)
-                    }
-                    Event::LostPeer(peer_id) => {
-                        OP_COUNTERS.inc("smp.event.lost_peer");
-                        lost_peer(&peer_info, peer_id);
-                        notify_subscribers(
-                            SharedMempoolNotification::PeerStateChange,
-                            &subscribers,
-                        );
-                        future::ready(None)
-                    }
-                    // Pass through messages to next combinator
-                    Event::Message((peer_id, msg)) => future::ready(Some((peer_id, msg))),
-                    _ => {
-                        security_log(SecurityEvent::InvalidNetworkEventMP)
-                            .error("UnexpectedNetworkEvent")
-                            .data(&network_event)
-                            .log();
-                        unreachable!("Unexpected network event")
+    let mut workers_available = smp.config.shared_mempool_max_concurrent_inbound_syncs;
+    let (sender, mut receiver) = mpsc::channel(workers_available);
```

Comment:
Looks good! Maybe a comment like "Use a futures mpsc channel to implement a semaphore, where the bounded channel contains all available work permits. When a worker wishes to handle a smp message, it must first acquire one of the work permits (and potentially block until one is available) before spawning its task. Because the channel is bounded, only `workers_available` number of workers may execute concurrently." Also rename `sender` + `receiver` to like `available_work_permits_tx` + `available_work_permits_rx` or something (doesn't have to be this specifically).